widgets/src/input/visual_text.rs

```diff
@@ -0,0 +1,119 @@
+    text_glyph(self.substr(..), &style, TextAlign::Start, clamp.max)
+  }
+
+  fn paint(&self, painter: &mut Painter, style: PaintingStyle, glyphs: &VisualGlyphs, rect: Rect) {
+    paint_text(painter, glyphs, style, rect);
+  }
+}
+
+/// The [`VisualTextGlyphs`] will give Provider of Stateful<TextGlyphCache> to
+/// the descendants Widgets, and can attach [`TextGlyphsPainter`] as child to
+/// paint the text
+#[derive(Declare)]
+pub struct VisualTextGlyphs<T>
+where
+  T: VisualText + Clone + 'static,
+{
+  pub text: T,
+}
+
+#[derive(SingleChild, Declare)]
+struct LayoutGlyphs<T>
+where
+  T: VisualText + Clone + 'static,
+{
+  text: T,
+}
+
+impl<T: 'static + VisualText + Clone> Render for LayoutGlyphs<T> {
+  fn perform_layout(&self, clamp: BoxClamp, ctx: &mut LayoutCtx) -> Size {
+    let cache = Provider::of::<Stateful<TextGlyphCache<T>>>(ctx).unwrap();
+
+    if cache
+      .read()
+      .text_glyph()
+      .as_ref()
+      .is_none_or(|c| c.text() != &self.text)
+    {
+      let glyphs = self.text.layout_glyphs(clamp, ctx);
+      *cache.write() = TextGlyphCache::new(self.text.clone(), glyphs);
+    }
+
+    let size = cache
+      .read()
+      .text_glyph()
+      .as_ref()
+      .map(|t| t.box_rect().size)
+      .unwrap_or_default();
+
+    let child = ctx.assert_single_child();
+    let child_size = ctx.perform_child_layout(child, clamp);
+
+    clamp.clamp(size.max(child_size))
+  }
+}
+
+/// The TextGlyphsPainter Widget will paint the glyphs get from the Provider of
+/// Stateful<TextGlyphCache<T>>, should be use as child of VisualTextGlyphs
+pub struct TextGlyphsPainter<T> {
+  marker: std::marker::PhantomData<T>,
+}
+
+impl<T> TextGlyphsPainter<T> {
+  pub fn new() -> Self { TextGlyphsPainter { marker: std::marker::PhantomData } }
+}
+
+impl<T: 'static + VisualText> Render for TextGlyphsPainter<T> {
+  fn paint(&self, ctx: &mut PaintingCtx) {
+    let box_rect = Rect::from_size(ctx.box_size().unwrap());
+    if let Some(rect) = ctx.painter().intersection_paint_bounds(&box_rect) {
+      Provider::of::<Stateful<TextGlyphCache<T>>>(ctx)
+        .map(|t| t.clone_reader())
+        .and_then(|p| {
+          p.read().text_glyph().as_ref().map(|c| {
+            let style = Provider::of::<PaintingStyle>(ctx).map(|p| p.clone());
+            c.paint(ctx.painter(), style.unwrap_or(PaintingStyle::Fill), rect)
+          })
+        });
+    };
+  }
+
+  fn perform_layout(&self, clamp: BoxClamp, _ctx: &mut LayoutCtx) -> Size { clamp.max }
+
+  fn hit_test(&self, _: &mut HitTestCtx, _: Point) -> HitTest {
+    HitTest { can_hit_child: false, hit: false }
+  }
+}
+
+impl<'w, T: VisualText + 'static + Clone> ComposeChild<'w> for VisualTextGlyphs<T> {
+  type Child = Vec<Widget<'w>>;
+  fn compose_child(this: impl StateWriter<Value = Self>, children: Self::Child) -> Widget<'w> {
+    fn_widget! {
+      Providers::new([Provider::new(Stateful::new(TextGlyphCache::<T>::default()))])
```

Comment:
Now, you can create a provider using `Provider::value_of_writer` to allow both the `Stateful` and write references to be obtained.